controller/contacts.js

```diff
@@ -0,0 +1,127 @@
+const service = require("../service");
+const { patternContact, patternFavorite } = require("../joi");
+
+const get = async (req, res, next) => {
+  try {
+    const contacts = await service.getAllContacts();
+    res.json({ data: contacts, status: "success", code: 200 });
+  } catch (err) {
+    next(err);
+  }
+};
+
+const getById = async (req, res, next) => {
+  const { contactId } = req.params;
+  try {
+    const contact = await service.getContactById(contactId);
+    res.json({ data: contact, status: "success", code: 200 });
+  } catch (err) {
+    next(res.json({ message: "Not found", status: "error", code: 404 }));
+  }
+};
+
+const add = async (req, res, next) => {
+  const validated = patternContact.validate(req.body);
+  if (validated.error) {
+    res.json({ message: validated.error.message, status: "error", code: 400 });
+  }
+  const { name, email, phone } = validated.value;
+
+  if (name && email && phone) {
+    try {
+      const contact = await service.addContact({ name, email, phone });
+      res.json({ data: contact, status: "success", code: 201 });
+    } catch (err) {
+      next(err);
+    }
+  } else {
+    return res.json({
+      message: "missing required name field",
+      status: "error",
+      code: 400,
+    });
+  }
+};
+
+const remove = async (req, res, next) => {
+  const { contactId } = req.params;
+  try {
+    const contact = await service.removeContact(contactId);
+    if (contact !== null) {
+      res.json({
+        data: contact,
+        message: "contact deleted",
+        status: "success",
+        code: 200,
+      });
+    } else {
+      res.json({ message: "Not found", status: "error", code: 404 });
+    }
+  } catch (err) {
+    next(err);
+  }
+};
+
+const updateContact = async (req, res, next) => {
+  const { contactId } = req.params;
+  const validated = patternContact.validate(req.body);
+  if (validated.error) {
+    res.json({ message: validated.error.message, status: "error", code: 400 });
+  }
+  const { name, email, phone } = validated.value;
+
+  if (!name && !email && !phone) {
+    res.json({ message: "missing fields", status: "error", code: 400 });
```

Comment:
Czy to nie wyjdzie w 67 linijce?

Reply:
To samo co wyżej.

Reply:
Tutaj akurat dla put body nie przyjmuje jedynie sytuacji kiedy wszystkie trzy klucze są puste, nie da sie tego ustawić bezpośrednio w Joi dlatego zrobiłem osobny obiekt walidacji dla put i dla add, natomiast if moim zdaniem powinien zostać.
Popraw mnie jeśli się mylę ;D

Reply:
Wtrącę się i zostawię link do podobnego problemu na stack-u, powinno rozwiązać problem a lepiej walidację trzymać w całości w jednym miejscu jeśli to możliwe
https://stackoverflow.com/questions/43103400/is-it-possible-to-require-at-least-one-field-from-a-set-of-defined-fields